spec/ParseQuery.Aggregate.spec.js

```diff
@@ -0,0 +1,343 @@
+'use strict';
+const Parse = require('parse/node');
+const rp = require('request-promise');
+
+const masterKeyHeaders = {
+  'X-Parse-Application-Id': 'test',
+  'X-Parse-Rest-API-Key': 'test',
+  'X-Parse-Master-Key': 'test'
+}
+
+const masterKeyOptions = {
+  headers: masterKeyHeaders,
+  json: true
+}
+
+const loadTestData = () => {
+  const data1 = {score: 10, name: 'foo', sender: {group: 'A'}, size: ['S', 'M']};
+  const data2 = {score: 10, name: 'foo', sender: {group: 'A'}, size: ['M', 'L']};
+  const data3 = {score: 10, name: 'bar', sender: {group: 'B'}, size: ['S']};
+  const data4 = {score: 20, name: 'dpl', sender: {group: 'B'}, size: ['S']};
+  const obj1 = new TestObject(data1);
+  const obj2 = new TestObject(data2);
+  const obj3 = new TestObject(data3);
+  const obj4 = new TestObject(data4);
+  return Parse.Object.saveAll([obj1, obj2, obj3, obj4]);
+}
+
+describe('Parse.Query Aggregate testing', () => {
+  beforeEach((done) => {
+    loadTestData().then(done, done);
+  });
+
+  it('should only query aggregate with master key', (done) => {
+    Parse._request('GET', `aggregate/someClass`, {})
+      .then(() => {}, (error) => {
+        expect(error.message).toEqual('unauthorized: master key is required');
+        done();
+      });
+  });
+
+  it('invalid query', (done) => {
+    const options = Object.assign({}, masterKeyOptions, {
+      body: {
+        unknown: {},
+      }
+    });
+    rp.get(Parse.serverURL + '/aggregate/TestObject', options)
+      .catch((error) => {
+        expect(error.error.code).toEqual(Parse.Error.INVALID_QUERY);
+        done();
+      });
+  });
+
+  it('group by field', (done) => {
+    const options = Object.assign({}, masterKeyOptions, {
+      body: {
+        group: { _id: '$name' },
+      }
+    });
+    rp.get(Parse.serverURL + '/aggregate/TestObject', options)
+      .then((resp) => {
+        expect(resp.results.length).toBe(3);
+        done();
+      }).catch(done.fail);
+  });
+
+  it('group sum query', (done) => {
+    const options = Object.assign({}, masterKeyOptions, {
+      body: {
+        group: { _id: null, total: { $sum: '$score' } },
```

Comment:
_id should be transformed (from objectId), which would be true for all keys passed here :)

Reply:
So `group: { objectId: null, total: { $sum: '$score' } }`

And it should return ` [ { objectId: null, total: ... } ]` instead of `[ { _id: null, total: ... } ]`

So `group: { objectId: '$name', total: { $sum: '$score' } }`

Should return ` [ { name: ..., total: ... } ]`

Reply:
Yep, in and out, those should not be the internal storage columns but the externals

Reply:
Should an error be thrown if _id is used?

Reply:
Yep, most probably!